lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -748,28 +750,45 @@ def __init__(
         # instead of "unable to read rule dir /loki/rules/fake: no such file or directory"
         self.container.make_dir(self._rules_dir, make_parents=True)
 
-        events = self.charm.on[relation_name]
+        events = self._charm.on[relation_name]
+        self.framework.observe(self._charm.on.upgrade_charm, self._on_logging_relation_changed)
         self.framework.observe(events.relation_changed, self._on_logging_relation_changed)
         self.framework.observe(events.relation_departed, self._on_logging_relation_departed)
 
     def _on_logging_relation_changed(self, event):
         """Handle changes in related consumers.
 
+        Anytime there are changes in the relation between Loki
+        and its consumers charms.
+
+        Args:
+            event: a `CharmEvent` in response to which the consumer
+                charm must update its relation data.
+        """
+        if isinstance(event, RelationEvent):
+            self._process_logging_relation_changed(event.relation)
+        else:
+            # Upgrade event or other charm-level event
+            for relation in self._charm.model.relations[self._relation_name]:
+                self._process_logging_relation_changed(relation)
+
+    def _process_logging_relation_changed(self, relation: Relation):
+        """Handle changes in related consumers.
+
         Anytime there are changes in relations between Loki
         and its consumers charms, Loki set the `loki_push_api`
         into the relation data.
         Besides Loki generates alert rules files based what
         consumer charms forwards,
 
         Args:
-            event: a `CharmEvent` in response to which the Loki
-                charm must update its relation data.
+            relation: the `Relation` instance to update.
         """
-        if event.relation.data[self.charm.unit].get("data") is None:
-            event.relation.data[self.charm.unit].update({"data": self._loki_push_api})
+        if self._charm.unit.is_leader():
+            relation.data[self._charm.app].update({"loki_push_api": self._loki_push_api})
             logger.debug("Saving Loki url in relation data %s", self._loki_push_api)
 
-        if event.relation.data.get(event.relation.app).get("alert_rules") is not None:
+        if relation.data.get(relation.app).get("alert_rules"):
```

Comment:
Do we need to tell loki to reload?

Reply:
As far as I understand there is no need to reload Loki: https://grafana.com/docs/loki/latest/rules/